contracts/vault.gno

```diff
@@ -0,0 +1,223 @@
+// @param amount0 string the amount of token0 to deposit
+// @param amount1 string the amount of token1 to deposit
+// @param amount0Min string the minimum amount of token0 to deposit
+// @param amount1Min string the minimum amount of token1 to deposit
+// @param deadline int64 the deadline for the transaction
+func DepositMulti(tokenId uint64, amount0 string, amount1 string, amount0Min string, amount1Min string, deadline int64) {
+	vault := mustGetVault(tokenId)
+	assertVaultPaused(vault)
+
+	caller := std.PrevRealm().Addr()
+
+	// Pull tokens from the caller
+	transferFromByRegisterCall(vault.token0, caller, std.CurrentRealm().Addr(), amount0)
+	transferFromByRegisterCall(vault.token1, caller, std.CurrentRealm().Addr(), amount1)
+
+	// Approve the pool contract to spend the tokens
+	poolAddr := consts.POOL_ADDR
+	approveByRegisterCall(vault.token0, poolAddr, amount0)
+	approveByRegisterCall(vault.token1, poolAddr, amount1)
+
+	_, liquidity, amount0Used, amount1Used, _ := position.IncreaseLiquidity(tokenId, amount0, amount1, amount0Min, amount1Min, deadline)
+
+	increaseBalance(vault, caller.String(), liquidity)
+
+	// Send back any leftover tokens
+	amount0Parsed := subAmounts(amount0, amount0Used)
+	amount1Parsed := subAmounts(amount1, amount1Used)
+
+	if amount0Parsed.Cmp(uint256.NewUint(0)) > 0 {
+		transferByRegisterCall(vault.token0, caller, amount0Parsed.ToString())
+	}
+	if amount1Parsed.Cmp(uint256.NewUint(0)) > 0 {
+		transferByRegisterCall(vault.token1, caller, amount1Parsed.ToString())
+	}
+
+	std.Emit("Deposit", caller.String(), strconv.FormatUint(tokenId, 10), amount0Parsed.ToString(), amount1Parsed.ToString(), liquidity)
+}
+
+// DepositSingle deposits a single token into a vault
+// @param tokenId uint64 the vault id
+// @param token string the token to deposit
+// @param amount string the amount of token to deposit
+func DepositSingle(tokenId uint64, token string, amount string) {
+	vault := mustGetVault(tokenId)
+	assertVaultPaused(vault)
+
+	// TODO swap token to token0 or token1 if needed
+}
+
+// Withdraw withdraws tokens from a vault
+// @param tokenId uint64 the vault id
+// @param amount string the amount of liquidity to withdraw
+//
+// panic if the amount is greater than the caller's balance
+func Withdraw(tokenId uint64, amount string) {
+	vault := mustGetVault(tokenId)
+	caller := std.PrevRealm().Addr()
+	balance := mustGetBalance(vault, caller.String())
+	parsedAmount, err := uint256.FromDecimal(amount)
+	if err != nil {
+		panic(err)
+	}
+	if balance.Cmp(parsedAmount) < 0 {
+		panic("insufficient balance")
+	}
+
+	liquidityRatio := computeLiquidityRatio(tokenId, parsedAmount)
+	_, liquidity, _, _, amount0, amount1, _ := position.DecreaseLiquidity(tokenId, liquidityRatio, uint256.NewUint(0).ToString(), uint256.NewUint(0).ToString(), time.Now().Unix()+1, false)
+
+	// Transfer tokens to the caller
+	transferByRegisterCall(vault.token0, caller, amount0)
+	transferByRegisterCall(vault.token1, caller, amount1)
```

Comment:
these tokens are  looks like transferred to the caller before updating the user's balance.